core/src/display_object/movie_clip.rs

```diff
@@ -1151,119 +1723,77 @@ impl<'gc> TDisplayObject<'gc> for MovieClip<'gc> {
         &self,
         context: &mut UpdateContext<'_, 'gc, '_>,
         display_object: DisplayObject<'gc>,
-        init_object: Option<Object<'gc>>,
-        instantiated_from_avm: bool,
+        init_object: Option<Avm1Object<'gc>>,
+        instantiated_by: Instantiator,
         run_frame: bool,
     ) {
         self.set_default_instance_name(context);
 
-        if self.0.read().object.is_none() {
-            let version = context.swf.version();
-            let globals = context.avm1.global_object_cell();
+        let movie = self.movie().unwrap();
+        let library = context.library.library_for_movie_mut(movie.clone());
 
-            // If we are running within the AVM, this must be an immediate action.
-            // If we are not, then this must be queued to be ran first-thing
-            if instantiated_from_avm && self.0.read().avm1_constructor.is_some() {
-                let mut activation = Activation::from_nothing(
-                    context.reborrow(),
-                    ActivationIdentifier::root("[Construct]"),
-                    version,
-                    globals,
-                    (*self).into(),
-                );
+        // Attempt to divine the VM we should initialize this instance as.
+        // If our movie doesn't already have that determined, then this is the
+        // root movie clip and we need to scan the SWF for file attributes.
+        let tendency = library.vm_tendency().unwrap_or_else(|| {
```

Comment:
I feel that reading the FileAttributes tag and setting the AVM version should be done before post_instantiation, not here; it could be done before the call to post_instantiation inside `Player::set_root_movie` (maybe as a separate method of MovieClip). Similarly this would occur for `Loaders` before post_instantiation whenever we implement those. That is, FileAttributes is basically an extension of the SWF header and should be treated as such; we should make sure we read it before we do any sort of execution.

1) FileAttributes check can be skipped for SWF version 8 and earlier (guaranteed AVM1).
2) You don't need to `decode_tags`, you should only have to read a single tag as FileAttributes is guaranteed to be the first tag in an SWF. I've also tested this, and a FileAttributes tag that is not the very first tag is ignored, and the SWF defaults to AVM1.
3) Would be good to make `avm_version` a property of `SwfMovie` as well.

Reply:
The other comment to make `AvmType` non-optional on each movie's library gave me the opportunity to stuff the VM check in `library_for_movie_mut`. I also altered all the movie loader paths to always call this before doing anything else, so the AS3 check runs as early as possible.

I'm a bit wary of adding further properties to `SwfMovie` as it's currently immutable, and we already have `MovieLibrary` to store mutable properties relating to a loaded movie. Perhaps in a future PR we could merge these together and make (some parts of) `SwfMovie` mutable.